docs/01-app/02-guides/content-security-policy.mdx

```diff
@@ -257,6 +341,561 @@ export default async function Page() {
+
+### Limitations of SRI
+
+- **Experimental**: Feature may change or be removed
+- **Webpack only**: Not available with Turbopack
+- **App Router only**: Not supported in Pages Router
+- **Build-time only**: Cannot handle dynamically generated scripts
+
+## Development vs Production Considerations
+
+CSP implementation differs between development and production environments:
+
+### Development Environment
+
+In development, you may need to allow `'unsafe-eval'` for Hot Module Replacement (HMR):
+
+```ts filename="middleware.ts" switcher
+export function middleware(request: NextRequest) {
+  const nonce = Buffer.from(crypto.randomUUID()).toString('base64')
+  const isDev = process.env.NODE_ENV === 'development'
+  
+  const cspHeader = `
+    default-src 'self';
+    script-src 'self' 'nonce-${nonce}' 'strict-dynamic' ${isDev ? "'unsafe-eval'" : ''};
+    style-src 'self' 'nonce-${nonce}' ${isDev ? "'unsafe-inline'" : ''};
+    img-src 'self' blob: data:;
+    font-src 'self';
+    object-src 'none';
+    base-uri 'self';
+    form-action 'self';
+    frame-ancestors 'none';
+    upgrade-insecure-requests;
+`
+  
+  // Rest of middleware implementation
+}
+```
+
+```js filename="middleware.js" switcher
+export function middleware(request) {
+  const nonce = Buffer.from(crypto.randomUUID()).toString('base64')
+  const isDev = process.env.NODE_ENV === 'development'
+  
+  const cspHeader = `
+    default-src 'self';
+    script-src 'self' 'nonce-${nonce}' 'strict-dynamic' ${isDev ? "'unsafe-eval'" : ''};
+    style-src 'self' 'nonce-${nonce}' ${isDev ? "'unsafe-inline'" : ''};
+    img-src 'self' blob: data:;
+    font-src 'self';
+    object-src 'none';
+    base-uri 'self';
+    form-action 'self';
+    frame-ancestors 'none';
+    upgrade-insecure-requests;
+`
+  
+  // Rest of middleware implementation
+}
+```
+
+### Production Deployment
+
+Common issues in production:
+
+- **Nonce not applied**: Ensure your middleware runs on all necessary routes
+- **Static assets blocked**: Verify your CSP allows Next.js static assets
+- **Third-party scripts**: Add necessary domains to your CSP policy
+
+## Troubleshooting
+
+### Error Pages and CSP
+
+Static error pages (404, 500) may not work with strict CSP because they don't go through your middleware. Solutions:
+
+<AppOnly>
+
+1. **Custom error pages**: Create custom error pages that support dynamic rendering:
+
+```tsx filename="app/not-found.tsx" switcher
+export const dynamic = 'force-dynamic'
+
+export default function NotFound() {
+  return (
+    <div>
+      <h2>Not Found</h2>
+      <p>Could not find requested resource</p>
+    </div>
+  )
+}
+```
+
+```jsx filename="app/not-found.jsx" switcher
+export const dynamic = 'force-dynamic'
+
+export default function NotFound() {
+  return (
+    <div>
+      <h2>Not Found</h2>
+      <p>Could not find requested resource</p>
+    </div>
+  )
+}
+```
+
+2. **Error boundary**: Use error boundaries to handle CSP violations gracefully
```

Comment:
How would this look like?

Reply:
There's an event `securitypolicyviolation` - but how does it link to `Error Pages and CSP`, with error boundaries 🤔